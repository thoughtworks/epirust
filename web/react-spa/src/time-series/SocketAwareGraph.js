import React, { useState, useEffect } from 'react';
import Graph from './LineGraph';
import PropTypes from 'prop-types'

export default function SocketAwareGraph({ socket, simulationId, transformFn }) {
    const [dataBuffer, setDataBuffer] = useState([]);
    const [simulationEnded, setSimulationEnded] = useState(false);
    useEffect(() => {
        if (!socket) {
            setDataBuffer([])
            return
        }

        let buff = [];
        socket.emit('simulation_id', simulationId);

        socket.on('epidemicStats', function (messageRaw) {
            const message = messageRaw;

            let simulationEndedTemp = false
            if ("simulation_ended" in message) {
                simulationEndedTemp = true;
                socket.close();
            }
            else {
                const perHourStats = transformFn(message);
                buff.push(perHourStats);
            }
            if (message.hour % 100 === 0 || simulationEndedTemp) {
                setDataBuffer(buffer => {
                    let total = [...buffer, ...buff]
                    buff = [];
                    return total;
                });
            }
            if (simulationEndedTemp) {
                setSimulationEnded(true)
            }
        });
    }, [socket, simulationId, transformFn]);

    return (
        <Graph
            dataBuffer={dataBuffer}
            enableExport={simulationEnded}
            labels={["hour", "susceptible", "infected", "quarantined", "recovered", "deceased"]}
        />
    )
}

SocketAwareGraph.propTypes = {
    socket: PropTypes.object,
    simulationId: PropTypes.number.isRequired,
    transformFn: PropTypes.func.isRequired
};
