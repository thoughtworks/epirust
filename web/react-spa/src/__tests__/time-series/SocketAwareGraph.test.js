/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from 'react'
import SocketAwareGraph from '../../time-series/SocketAwareGraph'
import { render } from '@testing-library/react'
import MockSocket from 'socket.io-mock'
import Dygraph from 'dygraphs'
import { act } from 'react-dom/test-utils'

jest.mock('dygraphs')
jest.useFakeTimers();

const simulationId = 1231231231

test('should render SocketAwareGraph', () => {
    const { asFragment } = render(<SocketAwareGraph simulationId={simulationId} socket={null} />)
    expect(asFragment()).toMatchSnapshot()
})

test('should recieve data sent on socket and parse & set graph to empty', () => {
    const mockDygraphfn = Dygraph.mockImplementation(() => { })
    let socket = new MockSocket()
    const hourStatistics = {
        hour: 10,
        susceptible: 9,
        infected: 2,
        quarantined: 1,
        recovered: 0,
        deceased: 0
    }

    render(<SocketAwareGraph socket={socket.socketClient} simulationId={simulationId} />)

    act(() => {
        socket.emit("epidemicStats", hourStatistics)
    })
    expect(mockDygraphfn).toHaveBeenCalledTimes(0)
    jest.clearAllMocks()
})

test('should set dataBuffer and render graph and plot graph', () => {
    const mockDygraphfn = Dygraph.mockImplementation(() => { })
    let socket = new MockSocket()

    const hourStatistics = {
        hour: 100,
        susceptible: 9,
        infected: 2,
        quarantined: 1,
        recovered: 0,
        deceased: 0
    }

    render(<SocketAwareGraph socket={socket.socketClient} simulationId={simulationId} />)

    act(() => {
        socket.emit("epidemicStats", hourStatistics)
        jest.runAllTimers();
    })

    expect(mockDygraphfn).toHaveBeenCalledWith(expect.anything(), [Object.values(hourStatistics)], expect.anything())
    jest.clearAllMocks()
})


test('should set residue also into data buffer when simulation ended flag is true', () => {
    const updateSpyFn = jest.fn()
    const mockDygraphfn = Dygraph.mockImplementation(() => ({
        updateOptions: updateSpyFn
    }))
    let socket = new MockSocket()
    socket.socketClient.close = () => { }
    const hourStatistics = {
        hour: 100,
        susceptible: 9,
        infected: 2,
        quarantined: 1,
        recovered: 0,
        deceased: 0
    }
    const hourStatistics101 = { ...hourStatistics, hour: 101 }

    render(<SocketAwareGraph socket={socket.socketClient} simulationId={simulationId} />)

    act(() => {
        socket.emit("epidemicStats", hourStatistics)
        socket.emit("epidemicStats", hourStatistics101)
        socket.emit("epidemicStats", { "simulation_ended": true })
        jest.runAllTimers();
    })

    expect(mockDygraphfn).toHaveBeenCalledTimes(1)
    expect(updateSpyFn).toHaveBeenCalledWith({ file: [Object.values(hourStatistics), Object.values(hourStatistics101)] })
})

test("should enable export in graph if simulation has ended", () => {
    let socket = new MockSocket()
    socket.socketClient.close = () => { };
    const { container } = render(<SocketAwareGraph socket={socket.socketClient} simulationId={simulationId} />)
    expect(container.querySelector(".graph-actions .btn-secondary")).toBeDisabled()

    act(() => {
        socket.emit("epidemicStats", { "simulation_ended": true })
        jest.runAllTimers();
    })

    expect(container.querySelector(".graph-actions .btn-secondary")).toBeEnabled()
})

test("should close the socket on receiving simulation ended message", () => {
    let socket = new MockSocket();
    let closeCall = 0;
    socket.socketClient.close = () => { closeCall += 1 };
    render(<SocketAwareGraph socket={socket.socketClient} simulationId={simulationId} />);

    act(() => {
        socket.emit("epidemicStats", { "simulation_ended": true });
        jest.runAllTimers();
    })

    expect(closeCall).toBe(1);
});
