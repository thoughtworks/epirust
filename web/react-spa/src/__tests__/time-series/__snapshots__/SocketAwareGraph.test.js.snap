// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SocketAwareGraph 1`] = `
<DocumentFragment>
  <div
    class="graph-wrap"
  >
    <div
      data-testid="visualization"
      id="vis"
    />
    <div
      class="graph-actions"
    >
      <p
        class="action"
      >
        Zoom: 
        <span
          class="control"
        >
          click-drag
        </span>
      </p>
      <p
        class="action"
      >
        Pan: 
        <span
          class="control"
        >
          shift-click-drag
        </span>
      </p>
      <p
        class="action"
      >
        Restore: 
        <span
          class="control"
        >
          double-click
        </span>
      </p>
      <button
        class="btn btn-secondary"
        disabled=""
      >
        Export graph
      </button>
    </div>
  </div>
</DocumentFragment>
`;
